### PR TITLE
[source-mongodb] add subtype when converting binary

### PR DESCRIPTION
**Pull request by xiaohansong**
## What
https://github.com/airbytehq/airbyte/issues/40525

for binary typed `_id`, we first base64 encode them into a readable string, and decode it back to binary and send to mongo as a cursor anchor.

When decoding it back to binary, we should also take binary subtype into consideration, otherwise it won't be converted back to the original value.

## How

In state message we will also need to track binSubtype. normally it should be null, unless the parent type is Binary. To migrate existing state where it does not have this new type, it will be defaulted to Binary(subtype 0) 

## Can this PR be safely reverted and rolled back?

- [ ] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Comment by evantahler** [comment]
Good find!

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/MongoDbInitialLoadRecordIterator.java [hunk 1]
```diff
@@ -128,7 +141,7 @@ private Bson buildFilter() {
             case OBJECT_ID -> new BsonObjectId(new ObjectId(state.id()));
             case INT -> new BsonInt32(Integer.parseInt(state.id()));
             case LONG -> new BsonInt64(Long.parseLong(state.id()));
-            case BINARY -> parseBinaryIdString(state.id());
+            case BINARY -> parseBinaryIdString(state.id(), state.binarySubType());
```
Comment (rodireich) [inline airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/MongoDbInitialLoadRecordIterator.java:144]: This is intended to ensure there's no multiple binary subtype of binary _id on a single collection.
Am I reading it correctly?

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/state/MongoDbStateManager.java [hunk 2]
```diff
@@ -290,28 +301,24 @@ public AirbyteStateMessage createFinalStateMessage(final ConfiguredAirbyteStream
     if (stream.getSyncMode() == INCREMENTAL) {
       final AirbyteStreamNameNamespacePair pair = new AirbyteStreamNameNamespacePair(stream.getStream().getName(), stream.getStream().getNamespace());
       if (!streamPairToLastIdMap.containsKey(pair)) {
+        // fix me!
```
Comment (rodireich) [inline airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/state/MongoDbStateManager.java:304]: ??

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/state/MongoDbStreamState.java [hunk 3]
```diff
@@ -4,4 +4,11 @@
 
 package io.airbyte.integrations.source.mongodb.state;
 
-public record MongoDbStreamState(String id, InitialSnapshotStatus status, IdType idType) {}
+public record MongoDbStreamState(String id, InitialSnapshotStatus status, IdType idType, Byte binarySubType) {
+
+  public MongoDbStreamState(String id, InitialSnapshotStatus status, IdType idType) {
+    this(id, status, idType, (byte) 0);
```
Comment (rodireich) [inline airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/state/MongoDbStreamState.java:10]: Will this allow for existing saved saved binary _id to transition in a non breaking way?
assuming this is subtype 0 (= generic binary).

What would happen to saved UUID (type 4)? 
Do we ensure they will also have a smooth transition?

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/MongoDbInitialLoadRecordIterator.java (hunk 1 above)
Comment (xiaohansong) [reply airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/MongoDbInitialLoadRecordIterator.java:144]: not really - this is to make sure we write the correct filter when querying mongo from record iterator

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/state/MongoDbStateManager.java (hunk 2 above)
Comment (xiaohansong) [reply airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/state/MongoDbStateManager.java:304]: sorry forgot to cleanup. was thinking about creating a map but ended up not using it

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/state/MongoDbStreamState.java (hunk 3 above)
Comment (xiaohansong) [reply airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/state/MongoDbStreamState.java:10]: not really - if idType is not binary, dev is not required to provide subType; that field will not be used.

existing state will be going through composeStateFromlastId to find the subtype.

**Review by rodireich — APPROVED** [review]

**Pull request merged**